crates/wasmtime/src/runtime/vm/traphandlers/backtrace.rs

```diff
@@ -360,31 +419,183 @@ impl<'a, T: 'static> CurrentActivationBacktrace<'a, T> {
+                    activation.exit_pc,
+                    activation.exit_fp,
+                    activation.entry_trampoline_fp,
+                )
             })
         };
 
-        CurrentActivationBacktrace { store, inner }
+        ActivationBacktrace { store, inner }
     }
 }
 
 #[cfg(feature = "debug")]
-impl<'a, T: 'static> Iterator for CurrentActivationBacktrace<'a, T> {
+impl<'a, T: 'static> Iterator for ActivationBacktrace<'a, T> {
     type Item = Frame;
     fn next(&mut self) -> Option<Self::Item> {
         self.inner.next()
     }
 }
+
+/// An iterator over all Wasm activations in a Store.
+#[cfg(feature = "debug")]
+pub(crate) struct StoreBacktrace<'a, T: 'static> {
+    /// The current activation iterator or the Store itself if no more
+    /// activations.
+    ///
+    /// This is `Option` so that we can move to the next while
+    /// transferring ownership of the Store without deconstructing
+    /// this whole iterator.
+    current: Option<StoreOrActivationBacktrace<'a, T>>,
+    activations: Vec<Activation>,
+}
+
+/// Either an iterator over a Wasm activation, or a `StoreContextMut`
+/// if no activations are left.
+#[cfg(feature = "debug")]
+enum StoreOrActivationBacktrace<'a, T: 'static> {
+    Store(StoreContextMut<'a, T>),
+    Activation(ActivationBacktrace<'a, T>),
+}
+
+#[cfg(feature = "debug")]
+impl<'a, T: 'static> StoreOrActivationBacktrace<'a, T> {
+    fn is_activation(&self) -> bool {
+        match self {
+            Self::Activation(_) => true,
+            _ => false,
+        }
+    }
+}
+
+#[cfg(feature = "debug")]
+impl<'a, T> StoreBacktrace<'a, T> {
+    /// Return an iterator over all Wasm activations in a Store, in
+    /// invocation order.
+    ///
+    /// The iterator captures the store with a mutable borrow, and
+    /// then yields it back at each frame. This ensures that the stack
+    /// remains live while still providing a mutable store that may be
+    /// needed to access items in the frame (e.g., to create new roots
+    /// when reading out GC refs).
+    ///
+    /// This serves as an alternative to `Backtrace::trace()` and
+    /// friends: it allows external iteration (and e.g. lazily walking
+    /// through frames in a stack) rather than visiting via a closure.
+    ///
+    /// # Safety
+    ///
+    /// Although the iterator provides mutable store as a public
+    /// field, this *must not* be used to mutate the stack itself that
+    /// this iterator is visiting. While the `store` technically owns
+    /// the stack in question, the only way to do this with the
+    /// current API would be to return back into the Wasm
+    /// activation. As long as this iterator is held and used while
+    /// within host code called from that activation (which will
+    /// ordinarily be ensured if the `store`'s lifetime came from the
+    /// host entry point) then everything will be sound.
+    pub(crate) unsafe fn new(store: StoreContextMut<'a, T>) -> StoreBacktrace<'a, T> {
```

Comment:
Kind of feel like this constructor should be safe and the `StoreBacktrace::store_mut` method should be unsafe, since that is the thing that actually lets you get the mutable store and potentially mutate the stack.

Reply:
Ah, that's a great point -- updated in 0e6ec66e1d.